internal/plugin.go

```diff
@@ -0,0 +1,660 @@
+		next func(context.Context, WorkerPluginStartWorkerOptions) error,
+	) error
+
+	// StopWorker is called to stop a worker. This is called on Worker.Stop or
+	// if Worker.Run is interrupted via its interrupt channel. However, if a
+	// fatal worker error occurs during Worker.Run, this may not be called.
+	// Implementers can account for this situation by setting OnFatalError in
+	// the worker options. Implementers should invoke next.
+	StopWorker(
+		ctx context.Context,
+		options WorkerPluginStopWorkerOptions,
+		next func(context.Context, WorkerPluginStopWorkerOptions),
+	)
+
+	// ConfigureWorkflowReplayer is called when the workflow replayer is created
+	// but before the options are validated. This call gives plugins a chance to
+	// adjust options as needed. This often includes adding interceptors.
+	ConfigureWorkflowReplayer(context.Context, WorkerPluginConfigureWorkflowReplayerOptions) error
+
+	// ReplayWorkflow is called for each individual workflow replay on the
+	// replayer. Implementers should return an error or invoke next.
+	ReplayWorkflow(
+		ctx context.Context,
+		options WorkerPluginReplayWorkflowOptions,
+		next func(context.Context, WorkerPluginReplayWorkflowOptions) error,
+	) error
+
+	// Plugins must embed [go.temporal.io/sdk/worker.PluginBase].
+	mustEmbedWorkerPluginBase()
+}
+
+// WorkerPluginBase must be embedded into worker plugin implementations.
+//
+// Exposed as: [go.temporal.io/sdk/worker.PluginBase]
+//
+// NOTE: Experimental
+type WorkerPluginBase struct{}
+
+var _ WorkerPlugin = struct {
+	pluginNamePanicForTypeChecking
+	WorkerPluginBase
+}{}
+
+// WorkerPluginConfigureWorkerOptions are options for ConfigureWorker on a
+// worker plugin.
+//
+// Exposed as: [go.temporal.io/sdk/worker.PluginConfigureWorkerOptions]
+//
+// NOTE: Experimental
+type WorkerPluginConfigureWorkerOptions struct {
+	// WorkerInstanceKey is the unique, immutable instance key for this worker.
+	WorkerInstanceKey string
+
+	// TaskQueue is the immutable task queue for this worker.
+	TaskQueue string
+
+	// WorkerOptions are the set of mutable options that can be adjusted by
+	// plugins.
+	WorkerOptions *WorkerOptions
+
+	// WorkerRegistryOptions are the set of callbacks that can be adjusted by
+	// plugins. If adjusting a callback that is already set, implementers may
+	// want to take care to invoke the existing callback inside their own.
+	WorkerRegistryOptions *WorkerPluginConfigureWorkerRegistryOptions
+}
+
+// WorkerPluginConfigureWorkerRegistryOptions are the set of callbacks that can
+// be adjusted by plugins when configuring workers. If adjusting a callback that
+// is already set, implementers may want to take care to invoke the existing
+// callback inside their own.
+//
+// Exposed as: [go.temporal.io/sdk/worker.PluginConfigureWorkerRegistryOptions]
+//
+// NOTE: Experimental
+type WorkerPluginConfigureWorkerRegistryOptions struct {
+	OnRegisterWorkflow        func(any, RegisterWorkflowOptions)
```

Comment:
A bit pedantic, but could we document these callbacks? They are a bit unclear since they have to take `any` as the first parameter